packages/data/src/types.ts

```diff
@@ -37,13 +43,187 @@ export interface ReduxStoreConfig<
+ * registered stores.
+ */
+type Store<
+	StoreRef extends Stores[ keyof Stores ] | keyof Stores
+> = StoreRef extends AnyConfig
+	? StoreRef
+	: StoreRef extends keyof Stores
+	? Stores[ StoreRef ]
+	: never;
+
+/**
+ * Removes the first argument from a function
+ *
+ * This is designed to remove the `state` parameter from
+ * registered selectors since that argument is supplied
+ * by the editor when calling `select(…)`.
+ *
+ * For functions with no arguments, which some selectors
+ * are free to define, returns the original function.
+ */
+type CurriedState< F > = F extends ( state: any, ...args: infer P ) => infer R
+	? ( ...args: P ) => R
+	: F;
+
+/**
+ * Returns a function whose return type is a Promise of the given return type.
+ *
+ * This is designed to take existing selectors and return a resolveSelector
+ * version of it which returns a Promise of the original return type. Promises
+ * flatten naturally and so if the original function already returned a Promise
+ * we can reuse the existing function type.
+ */
+type Resolvable<
+	F extends ( ...args: any[] ) => any
+> = ReturnType< F > extends Promise< any >
+	? F
+	: ( ...args: Parameters< F > ) => Promise< ReturnType< F > >;
+
+/**
+ * These fields are excluded from the resolveSelect mapping.
+ *
+ * @see mapResolveSelectors
+ */
+type NonResolveSelectFields =
+	| 'getIsResolving'
+	| 'hasStartedResolution'
+	| 'hasFinishedResolution'
+	| 'isResolving'
+	| 'getCachedResolvers';
+
 export interface DataRegistry {
-	register: ( store: StoreDescriptor< any > ) => void;
+	/** Apply multiple store updates without calling store listeners until all have finished */
+	batch( executor: () => void ): void;
+
+	/** Returns the available actions for a given store. */
+	dispatch< StoreRef extends Stores[ keyof Stores ] | keyof Stores >(
+		store: StoreRef
+	): NonNullable< Store< StoreRef >[ 'actions' ] >;
+
+	/** Registers a new store into the registry. */
+	register( store: StoreDescriptor< any > ): void;
+
+	/** Given a namespace key and store description, registers a new Redux store into the registry. */
+	registerStore( name: string, store: StoreDescriptor< any > ): ReduxStore;
+
+	/**
+	 * Returns a version of the available selectors for a given store that returns
+	 * a Promise which resolves after all associated resolvers have finished.
+	 */
+	resolveSelect<
+		StoreRef extends Stores[ keyof Stores ] | keyof Stores,
+		Selectors extends Store< StoreRef >[ 'selectors' ]
+	>(
+		store: StoreRef
+	): NonNullable<
+		{
+			[ Name in keyof Selectors as Exclude<
+				Name,
+				NonResolveSelectFields
+			> ]: Resolvable< CurriedState< Selectors[ Name ] > >;
+		}
+	>;
+
+	/** Returns the available selectors for a given store. */
+	select<
+		StoreRef extends Stores[ keyof Stores ] | keyof Stores,
+		Selectors extends Store< StoreRef >[ 'selectors' ]
+	>(
+		store: StoreRef
+	): NonNullable<
+		{
+			[ Name in keyof Selectors ]: CurriedState< Selectors[ Name ] >;
+		}
+	>;
```

Comment:
Something about the type inference here isn't exact:

![CleanShot 2022-03-09 at 11 16 54](https://user-images.githubusercontent.com/205419/157421640-87663c9b-7884-406d-b70a-711e19dfce18.png)

The error doesn't show up when I use the `getEntityRecord` selector directly without going through `select`.

I've had a similar problem in https://github.com/WordPress/gutenberg/pull/39025 and [ended up](https://github.com/WordPress/gutenberg/pull/39025/commits/25280cd5a799852a7f1bbc85753d61f19f16d61f#diff-801cf8c502406e9478ef173002e2477051c867463f3d9f0f9074626b51bb07fc) with one of these super long type declarations that test for every number of arguments separately:

```ts
type IsValidArg<T> = T extends object ? keyof T extends never ? false : true : true;
type ReplaceReturnType<T, TNewReturn> = T extends (a: infer A, b: infer B, c: infer C, d: infer D, e: infer E, f: infer F, g: infer G, h: infer H, i: infer I, j: infer J) => infer R ? (
    IsValidArg<J> extends true ? (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J) => TNewReturn :
    IsValidArg<I> extends true ? (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I) => TNewReturn :
    IsValidArg<H> extends true ? (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H) => TNewReturn :
    IsValidArg<G> extends true ? (a: A, b: B, c: C, d: D, e: E, f: F, g: G) => TNewReturn :
    IsValidArg<F> extends true ? (a: A, b: B, c: C, d: D, e: E, f: F) => TNewReturn :
    IsValidArg<E> extends true ? (a: A, b: B, c: C, d: D, e: E) => TNewReturn :
    IsValidArg<D> extends true ? (a: A, b: B, c: C, d: D) => TNewReturn :
    IsValidArg<C> extends true ? (a: A, b: B, c: C) => TNewReturn :
    IsValidArg<B> extends true ? (a: A, b: B) => TNewReturn :
    IsValidArg<A> extends true ? (a: A) => TNewReturn :
    () => TNewReturn
) : never
```

[Credit goes to the following SO answer.](https://stackoverflow.com/questions/50011616/typescript-change-function-type-so-that-it-returns-new-value)

Reply:
I think the error might be coming from the type for `rememo` there as I'm not seeing it. For my own testing I used this to get around having a real type for `rememo`, which reminds me, we should file a PR in DefinitelyTyped for it

```ts
const createSelector = rememoCreateSelect as <F>(selector: F, deps: any) => F;
```

https://user-images.githubusercontent.com/5431237/159084572-4fcca9ae-0dcb-47a3-bca1-fd5db856ac12.mov

you can also see in that screencast my original intended method for adding core styles into the type system. `typeof storeConfig`, or in this case, since `storeConfig` is a function that produces it, `ReturnType<typeof storeConfig>` so we don't even need to know about `ReduxStoreConfig` (though it's obviously fine to manually add `ReduxStoreConfig<…, …, …>`)

Reply:
@dmsnell Could it be that the [selector types signature PR](https://github.com/WordPress/gutenberg/pull/39025) now stores these signatures in a separate file and doesn't associate them with the selector functions? Because this changed between my original comment and your reply. I'm asking because I noticed all the arguments are of the type `any`:

![CleanShot 2022-03-20 at 20 52 26](https://user-images.githubusercontent.com/205419/159181729-87308aa6-4297-40bb-97df-048b5e99dd25.png)

And I hoped to see `kind: Kind, name: Name, key: KeyOf< EntityRecordType< Kind, Name > >` etc

Reply:
> Could it be that the https://github.com/WordPress/gutenberg/pull/39025 now stores these signatures in a separate file and doesn't associate them with the selector functions?

I'm not sure what you mean here. This code is looking at the selectors directly and should report exactly what TypeScript knows about those selectors.

If you create `ReduxStoreConfig<State, Actions, Selectors>` manually then it might behave differently because we're telling the type system to trust that what we're saying is more accurate than what it infers. I don't think I understand your question exactly though, at least I am pretty sure I'm missing the question you are asking.

> all the arguments are of the type any:

Since the selectors are written in JS we aren't getting types. This should match what info TypeScript has. As we add types or convert those modules to TypeScript the argument type information should appear automatically in this wrapped type. I'll try and get an example screencast of that.

Reply:
> I'll try and get an example screencast of that.

What I meant was that once you wire the TypeScript type for the getEntityRecord selector instead of using the vanilla JS version, the function signature will break as on the original screenshot I posted above.

The reason it does not break like that at the moment is that the TS signature and the JS selector got decoupled in my PR after I initially commented.

Reply:
I'm still not getting it, which may be because I just don't comprehend what's going on.

How does "a TS signature and JS selector [get] decoupled"?